<a href="https://colab.research.google.com/github/ggomes88/PythonFinance/blob/main/Clustering_IBRA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. DEFININDO O PROBLEMA**

O Objetivo deste estudo é encontrar ativos semelhantes dentro do índice IBRA da B3 e assim encontrar pares de ativos para uma estratégia de Pair Trading.

Pair trading é uma estratégia de arbritragem estatística na qual um par de ações é negociado sem levar em conta a tendência do mercado.

 O IBrA é o indicador do desempenho médio das cotações de todos os ativos negociados no mercado a vista (lote-padrão) da B3 que atendam a critérios mínimos de liquidez e presença em pregão, de forma a oferecer uma visão ampla do mercado acionário.

Para essa análise iremos utilizar dados obtivos pela API do Yahoo Finance e bibliotecas do Python voltadas para Machine LEarning e Estatística


#**2. Carregando as Bibliotecas do Python e os Dados da Yahoo Finance**

### **2.1 Carregando as bibliotecas**

In [4]:
# Carregando Bibliotecas do Python

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pandas import read_csv, set_option
from pandas.plotting import scatter_matrix
import seaborn as sns
from sklearn.preprocessing import StandardScaler
import datetime as dt
import pandas_datareader as dr
!pip install yfinance
import yfinance as yf

#Importando os Algoritmos de Machine LEarning
from sklearn.cluster import KMeans, AgglomerativeClustering,AffinityPropagation, DBSCAN
from scipy.cluster.hierarchy import fcluster
from scipy.cluster.hierarchy import dendrogram, linkage, cophenet
from scipy.spatial.distance import pdist
from sklearn.metrics import adjusted_mutual_info_score
from sklearn import cluster, covariance, manifold

#Outros pacotes
import matplotlib.ticker as ticker
from itertools import cycle

[*********************100%***********************]  178 of 178 completed


pandas.core.frame.DataFrame

### **2.2 Carregando os dados**
Para essa análise iremos utilizar o período de janeiro de 2018 a julho de 2021.
Utilizei um csv contendo todos os ativos do IBRA que baixei direto do site da B3 e converti para a nomenclatura utilizada pelo Yahoo Finance. Transformei esses dados em lista para poder utilizar como argumento para a API do Yahoo Finance. Após baixar os dados, criei um dataset contendo apenas os preços de fechamento dos ativos.

In [5]:
#Baixando lista de ativos integrantes do IBRA do CSV com a nomenclatura do Yahoo Finance

ibra = pd.read_csv('ibra.csv',header=None)

#Convertendo dataframe para lista
ibra = ibra[0].values.tolist()

#Baixando dados do yfinance
inicio=dt.datetime(2018,1,1)
fim=dt.datetime(2021,7,1)
df=yf.download(ibra,inicio,fim)

#Limpando o DF para usar apenas os preços fechamento do dia
dataset = df['Adj Close']
type(dataset)

[*********************100%***********************]  178 of 178 completed


pandas.core.frame.DataFrame

# 3. Análise Exploratória

178 ativos foram baixados do Yahoo Finance, com 866 dias de negociação:

In [7]:
dataset.shape

(866, 178)

Vemos que alguns ativos não possuem dados em algumas datas. Podem ser ativos que foram adicionados após 2018 ou ativos que não constam na base de dados

In [8]:
set_option('display.width', 100)
dataset.head(5)

AALR3.SA  ABCB4.SA  ABEV3.SA  ...  WEGE3.SA  WIZS3.SA  YDUQ3.SA
Date                                      ...                              
2018-01-02    14.740    13.700    19.845  ...     9.183     8.607    29.988
2018-01-03    14.810    13.758    19.890  ...     9.044     8.536    30.015
2018-01-04    14.938    13.716    19.781  ...     8.956     8.465    29.891
2018-01-05    14.899    13.915    19.854  ...     8.992     8.586    29.483
2018-01-08    14.641    14.047    19.817  ...     9.066     8.550    29.590

[5 rows x 178 columns]

Vamos analisar os dados estatísticos dos dados:

In [9]:
set_option('precision', 3)
dataset.describe()

AALR3.SA  ABCB4.SA  ABEV3.SA  ...  WEGE3.SA  WIZS3.SA  YDUQ3.SA
count   860.000   860.000   860.000  ...   860.000   860.000   860.000
mean     13.473    15.016    16.296  ...    18.727     8.433    29.871
std       3.138     2.106     2.532  ...    11.842     2.247     6.872
min       8.252    10.288    10.695  ...     7.300     5.062    18.309
25%      11.110    13.325    14.766  ...     9.006     6.355    25.701
50%      12.944    14.813    16.567  ...    11.805     8.447    28.976
75%      15.146    16.671    17.762  ...    31.859     9.735    32.917
max      22.960    22.335    22.276  ...    45.849    17.740    54.680

[8 rows x 178 columns]

#3.2 Visualização dos Dados